TwoFactorAuth/Observer/ControllerActionPredispatch.php

```diff
@@ -106,22 +135,40 @@ public function execute(Observer $observer)
         $fullActionName = $controllerAction->getRequest()->getFullActionName();
         $user = $this->getUser();
 
+        $this->tokenManager->readConfigToken();
+
         if (in_array($fullActionName, $this->tfa->getAllowedUrls(), true)) {
             //Actions that are used for 2FA must remain accessible.
             return;
         }
 
         if ($user) {
-            if ($this->configRequestManager->isConfigurationRequiredFor((int)$user->getId())) {
-                //User must configure 2FA first
-                $this->tokenManager->readConfigToken();
+            $configurationStillRequired = $this->configRequestManager->isConfigurationRequiredFor((int)$user->getId());
+            $toActivate = $this->tfa->getProvidersToActivate((int)$user->getId());
+            $toActivateCodes = [];
+            foreach ($toActivate as $toActivateProvider) {
+                $toActivateCodes[] = $toActivateProvider->getCode();
+            }
+            $currentlySkipped = $this->session->getData('tfa_skipped_config') ?? [];
+            $accessGranted = $this->tfaSession->isGranted();
+
+            if (!$accessGranted
+                && $configurationStillRequired
+                && array_diff($toActivateCodes, array_keys($currentlySkipped))
+            ) {
                 //User needs special link with a token to be allowed to configure 2FA
-                $this->redirect('tfa/tfa/requestconfig');
+                if ($this->authorization->isAllowed(Requestconfig::ADMIN_RESOURCE)) {
```

Comment:
It will but there are cases where it will causes redirect loops. For example, an unconfigured tfa site where a user without `Magento_TwoFactorAuth::config` tries to login it will ultimately get redirected to `requestconfig` which will redirect to `tfa/tfa/configure` but that requires `Magento_TwoFactorAuth::config` so it redirects to login->dashboard->requestconfig->tfa/tfa/index->tfa/configure->login->dashboard... etc. This check catches the loop and displays a friendly error. Same thing for the similar check below but for personal 2fa.

Reply:
However, this does make me realize i have the wrong acl rule being checked here. I will fix it, it should be `Magento_TwoFactorAuth::config` not `::tfa`